### DISCUSSION TITLE
Linux - ncdu with sudo

### DISCUSSION THREAD

**Discussion by Justinzobel** [Homebrew/discussions#5650] [Linux] [answered]
### Additional Info

- [ ] I am using Windows Subsystem for Linux

### Output of `brew config`

```shell
N/A
```

### Output of `brew doctor`

```shell
N/A
```

### Description of issue

How does one use sudo with ncdu to look at filesystem directories, e.g. `sudo ncdu /`?

**Comment by ZhongRuoyu** [comment, marked as answer]
If your problem is `sudo` complaining it cannot find `ncdu`, this is because `sudo` resets the `PATH`. Instead of `sudo ncdu /` you can do `` sudo `which ncdu` / ``.

**Reply by Justinzobel** [reply]
Thanks! For me on `fish` it was `sudo $(which ncdu) /`

**Reply by gromgit** [reply]
`sudo` only resets `PATH` _after_ it searches for the specified command, so substituting `which ncdu` should have no net effect.

@Justinzobel, what's the exact error you see with `sudo ncdu /`?

**Reply by Justinzobel** [reply]
Yeah it was a case of not in path.

**Reply by ZhongRuoyu** [reply]
> `sudo` only resets `PATH` _after_ it searches for the specified command

If `secure_path` is set in `sudoers`, then `sudo` only searches for commands from the paths set there. From [manpage](https://www.sudo.ws/docs/man/sudoers.man/):

> secure_path
> If set, `sudo` will use this value in place of the user's `PATH` environment variable. There are two basic use cases for secure_path:
>
> 1. To make it possible for `sudo` to find system administrator commands located in directories that may not be in the default user path, such as /usr/sbin.
> 2. To help protect scripts and programs that execute other commands without first setting `PATH` to a safe value. Otherwise, a user with limited privileges may be able to run arbitrary commands by manipulating the `PATH` if the command being run executes other commands without using a fully-qualified path name.
>
> Users in the group specified by the exempt_group option are not affected by secure_path. This option is disabled by default.


It's not set by default on macOS, but some distros, like Debian, [do this](https://salsa.debian.org/sudo-team/sudo/-/blob/debian/1.9.16-1/debian/etc/sudoers?ref_type=tags#L11):

```sudoers
Defaults	secure_path="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
```

I believe OP's `sudo ncdu` invocation failed because of this.
